jwst/datamodels/util.py

```diff
@@ -488,3 +493,157 @@ def get_envar_as_boolean(name, default=False):
+    else:
+        unit_model = open(unit_shape)
+
+    # Size of the primary array.
+    primary_array_name = unit_model.get_primary_array_name()
+    primary_array = getattr(unit_model, primary_array_name)
+    size = primary_array.nbytes
+    for dimension in shape:
+        size *= dimension
+
+    # Get available memory
+    available = get_available_memory(include_swap=include_swap)
+    log.debug(f'Model size {bytes2human(size)} available system memory {bytes2human(available)}')
+
+    if size > available:
+        log.warning(
+            f'Model {model_type} shape {shape} requires {bytes2human(size)} which is more than'
+            f' system available {bytes2human(available)}'
+        )
+
+    if allowed and size > (allowed * available):
+        log.debug(
+            f'Model size greater than allowed memory {bytes2human(allowed * available)}'
+        )
+        return False, size
+
+    return True, size
+
+
+def get_available_memory(include_swap=True):
+    """Retrieve available memory
+
+    Parameters
+    ----------
+    include_swap : bool
+        Include available swap in the calculation.
+
+    Returns
+    -------
+    available : number
+        The amount available.
+    """
+    system = platform_system()
+
+    # Apple MacOS
+    log.debug(f'Running OS is "{system}"')
+    if system in ['Darwin']:
+        return get_available_memory_darwin(include_swap=include_swap)
+
+    # Default to Linux-like:
+    return get_available_memory_linux(include_swap=include_swap)
+
+
+def get_available_memory_linux(include_swap=True):
+    """Get memory for a Linux system
+
+    Presume that the swap space as reported is accurate at the time of
+    the query and that any subsequent allocation will be held the value.
+
+    Parameters
+    ----------
+    include_swap : bool
+        Include available swap in the calculation.
+
+    Returns
+    -------
+    available : number
+        The amount available.
+    """
+    vm_stats = psutil.virtual_memory()
+    available = vm_stats.available
+    if include_swap:
+        swap = psutil.swap_memory()
+        available += swap.total
+    return available
+
+
+def get_available_memory_darwin(include_swap=True):
+    """Get the available memory on an Apple MacOS-like system
+
+    For Darwin, swap space is dynamic and will attempt to use the whole of the
+    boot partition.
+
+    If the system has been configured to use swap from other sources besides
+    the boot partition, that available space will not be included.
+
+    Parameters
+    ----------
+    include_swap : bool
+        Include available swap in the calculation.
+
+    Returns
+    -------
+    available : number
+        The amount available.
+    """
+    vm_stats = psutil.virtual_memory()
+    available = vm_stats.available
+    if include_swap:
+
+        # Attempt to determine amount of free disk space on the boot partition.
+        run_info = run(['/usr/sbin/bless', '--getBoot', '--plist'], capture_output=True)
```

Comment:
This is the result on my laptop (macOS 10.15.6):

```
$ /usr/sbin/bless --getBoot --plist
Can't access "efi-boot-device" NVRAM variable
```

Reply:
Go figure. Well, delving more around, it seems a more reliable method is using `/private/var/vm`. Will do that and avoid being "bless"ed.